walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long requestedDuration = request.getDuration();
+
+    if (Long.compare(requestedDuration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> occupiedTimeRanges =
+        getRelevantTimeRangesFromEvents(events, attendees);
+
+    return getAvailabeTimeRanges(occupiedTimeRanges, requestedDuration);
+  }
+
+  /**
+  * Finds the available {@code TimeRange} having at least a length of {@code requestedDuration},
+  * given some already occupied {@code TimeRange}. The resultant list of available
+  * {@code TimeRange} is sorted in ascending chronological order.
+  */
+  private List<TimeRange> getAvailabeTimeRanges(Collection<TimeRange> occupiedTimeRanges,
+      long requestedDuration) {
+    if (occupiedTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> availableTimeRanges = new LinkedList<>();
+    List<TimeRange> startOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                  TimeRange.ORDER_BY_START);
+    List<TimeRange> endOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                TimeRange.ORDER_BY_END);
+
+    // Potentially add the {@code TimeRange} before the first existing {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
+    ListIterator<TimeRange> startOrderedTimeRangesIterator = startOrderedTimeRanges.listIterator();
+    ListIterator<TimeRange> endOrderedTimeRangesIterator = endOrderedTimeRanges.listIterator();
+    TimeRange endingTimeRange = endOrderedTimeRangesIterator.next();
+    TimeRange startingTimeRange = startOrderedTimeRangesIterator.next();
+    while (true) {
+      // There is no available time after the current ending {@code TimeRange} and before the
+      // immediately following starting {@code TimeRange}.
+      if (endingTimeRange.overlaps(startingTimeRange)) {
+        if (startOrderedTimeRangesIterator.hasNext()) {
+          startingTimeRange = startOrderedTimeRangesIterator.next();
+        } else {
+          break;
+        }
+        continue;
```

Comment:
can just put this up in the second nested `if` after `startingTimeRange =` line since you're `break`ing in the `else` anyways.

Reply:
Yeah, I've considered this. The reason why I didn't do it is because I wanted to keep the logic for the iterator separate from the rest (the if-block). It's basically pointer++, but it looks long. I'll try to make this easier to read